src/main/kotlin/com/fasterxml/jackson/module/kotlin/MethodInstantiator.kt

```diff
@@ -0,0 +1,104 @@
+package com.fasterxml.jackson.module.kotlin
+
+import com.fasterxml.jackson.databind.DeserializationContext
+import com.fasterxml.jackson.databind.MapperFeature
+import com.fasterxml.jackson.module.kotlin.Instantiator.Companion.INT_PRIMITIVE_CLASS
+import java.lang.reflect.Method
+import kotlin.reflect.KFunction
+import kotlin.reflect.KParameter
+import kotlin.reflect.full.valueParameters
+
+// This class does not support inner constructor.
+internal class MethodInstantiator<T>(
+    kFunction: KFunction<T>,
+    private val method: Method,
+    private val instance: Any,
+    companionAccessible: Boolean
+) : Instantiator<T> {
+    override val hasInstanceParameter: Boolean = true
+    override val valueParameters: List<KParameter> = kFunction.valueParameters
+    private val accessible: Boolean = companionAccessible && method.isAccessible
+    private val bucketGenerator = BucketGenerator(valueParameters)
+
+    init {
+        method.isAccessible = true
+    }
+
+    private enum class ProcessingMode {
+        Instance, ParameterTypes, Mask, Marker
+    }
+
+    // argument size = parameterSize + maskSize + instanceSize(= 1) + markerSize(= 1)
+    private val defaultArgumentSize: Int by lazy {
+        valueParameters.size + bucketGenerator.maskSize + 2
+    }
+
+    // This initialization process is heavy and will not be done until it is needed.
+    private val localMethod: Method by lazy {
+        val instanceClazz = instance::class.java
+        val parameterTypes = method.parameterTypes
+
+        var processingMode = ProcessingMode.Instance
+        var innerIndex = 0
+
+        // argumentTypes = arrayOf(instanceClazz, *parameterTypes, *Array(bucketGenerator.maskSize) { INT_PRIMITIVE_CLASS }, Object::class.java)
+        val argumentTypes = Array(defaultArgumentSize) {
+            when(processingMode) {
+                ProcessingMode.Instance -> {
+                    processingMode = ProcessingMode.ParameterTypes
+                    instanceClazz
+                }
+                ProcessingMode.ParameterTypes -> {
+                    val next = parameterTypes[innerIndex]
+                    innerIndex++
+                    if (innerIndex == parameterTypes.size) {
+                        processingMode = ProcessingMode.Mask
+                        innerIndex = 0
+                    }
+
+                    next
+                }
+                ProcessingMode.Mask -> {
+                    innerIndex++
+                    if (innerIndex == bucketGenerator.maskSize) {
+                        processingMode = ProcessingMode.Marker
+                    }
+                    INT_PRIMITIVE_CLASS
+                }
+                ProcessingMode.Marker -> Object::class.java
+            }
+        }
```

Comment:
There are several processes that have become more complex as a result of optimization without using the `spread operator`.
Is it better to use `spread operator` to simplify processes with small cost that are called only once?

Reply:
If "called only once" means that during introspection needed for deserializers, definitely (and not every time deserialization occurs).

Reply:
> If "called only once" means that during introspection needed for deserializers, definitely (and not every time deserialization occurs).

That's what I wanted to convey.
Simplified.
04f8475